File: 973. K Closest Points to Origin/973. K Closest Points to Origin.py
from typing import List
import math
import heapq


class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        heap = []

        for p in points:
            distance = -(p[0] ** 2 + p[1] ** 2)

            if len(heap) < k:
                heapq.heappush(heap, (distance, p))
            else:
                heapq.heappushpop(heap, (distance, p))

        return [h[1] for h in heap]

File: 973. K Closest Points to Origin/973. K Closest Points to Origin.java
import java.util.PriorityQueue;

class Solution {

  public int[][] kClosest(int[][] points, int k) {
    PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) ->
      Double.compare(b[0] * b[0] + b[1] * b[1], a[0] * a[0] + a[1] * a[1])
    );

    for (int[] p : points) {
      heap.add(p);
      if (heap.size() > k) {
        heap.poll();
      }
    }

    int[][] ans = new int[k][2];
    for (int i = 0; i < k; i++) {
      ans[i] = heap.poll();
    }

    return ans;
  }
}
